### PR TITLE
ContainerCollection: add enrichers

### PR DESCRIPTION
**Pull request by alban**
# ContainerCollection: add enrichers

ContainerCollection: add enrichers

ContainerCollectionInitialize() now accepts functional options:

- WithDockerEnrichment(): enrich container metadata with the container
  name from the Docker API. Additionally, fetches initial containers
  from the Docker API.

- WithKubernetesEnrichment(): enrich container metadata from the
  Kubernetes API (Kubernetes namespace, pod name...). This enricher
  requires cgroups data to be prefilled.

- WithCgroupEnrichment(): enrich container metadata with cgroups paths
  and cgroup id. This enricher requires the pid to be prefilled.

- WithNamespaceEnrichment(): enrich container metadata with mount
  namespace id. This enricher requires the pid to be prefilled.

- WithRuncFanotify(): adds new containers from the runcfanotify package.
  This does not add any proper enricher by itself, but since the
  runcfanotify package does not directly support initial containers, it
  registers a hook on other functional options that support initial
  containers (e.g. WithDockerEnrichment) so it can detect when those
  containers terminate.

The implementation of enrichers in ContainerCollection allow to solve
the following problems:

1. Initial containers when using runcfanotify
2. Container termination when using runcfanotify

without adding interdependency between the different hook modes and the
enrichers. The Gadget Tracer Manager and the Local Gadget Manager don't
have to duplicate this code anymore. It is also simpler code:

* Gadget Tracer Manager
```
	opts := []containercollection.ContainerCollectionOption{
		containercollection.WithCgroupEnrichment(),
		containercollection.WithNamespaceEnrichment(),
		containercollection.WithKubernetesEnrichment(nodeName),
		containercollection.WithRuncFanotify(),
	}
	if withRuncFanotify {
		opts = append(opts, containercollection.WithRuncFanotify())
	}
	g.ContainerCollectionInitialize(opts...)
```

* Local Gadget Manager:
```
	l.ContainerCollection.ContainerCollectionInitialize(
		containercollection.WithCgroupEnrichment(),
		containercollection.WithNamespaceEnrichment(),
		containercollection.WithDockerEnrichment(),
		containercollection.WithRuncFanotify(),
	)
```

Additionally, the sync.Map now store pointers to ContainerDefinition as
previously discussed.

## How to use

[ describe what reviewers need to do in order to validate this PR ]

## Testing done

WIP

## TODO

- [x] Merge https://github.com/kinvolk/inspektor-gadget/pull/312 first
- [ ] Test with / without runcfanotify
- [ ] Test with / without podinformer

### CODE REVIEW

**Review by mauriciovasquezbernal — COMMENTED** [review]
I think it's a **great** steps towards a better and simpler code!  I left a lot of comments, most of them are just details. 

I'm also wondering if could change the pod informer to fit this model. I think we could implement a `WithK8sPodInformer` option taking into consideration the order how they are invoked. 

**Extending a bit the idea:**
The pod informer receives a pod from the k8s api server, this already has all the k8s information in (namespace, pod name, containers name, labels...) but some low level information about the container is missing (like pid, mount ns path, cgroup path). For this reason the PodInformer has a [criClient](https://github.com/kinvolk/inspektor-gadget/blob/38a7333a493cc9a3ff020b427c57a5a0c04428b6/pkg/gadgettracermanager/containerutils/containerutils.go#L74) to query the different container runtimes instance for the PID. I think in the case pod informer is used we could have a first enrichement step (`WithPidEnrichment`?) that fills the PID in and then others ones that need the pid like WithCgroupEnrichment, WithNamespaceEnrichment will work fine. 

I know that this could required a bit of rework but it'd improve the code a lot. Of course we can decide to do that in a different PR if it gets too complicated here. 

Finally, do you think it's easier to separate the change the uses a pointer instead of a structure of the container in the different places in its own commit?

File: pkg/runcfanotify/runcfanotify.go [hunk 1]
```diff
@@ -119,6 +119,41 @@ func cmdlineFromPid(pid int) []string {
...
+	defer unix.Close(int(pidfd))
+
+	for {
+		fds := []unix.PollFd{
+			{
+				Fd:      int32(pidfd),
+				Events:  unix.POLLIN,
+				Revents: 0,
+			},
+		}
```
[the remark this answers is not in the document]
Comment (alban) [reply pkg/runcfanotify/runcfanotify.go:144]: The man page says:

> The field revents is an output parameter, filled by the kernel with the events that actually occurred.

But it does not explicitly say that the kernel will initialise `revents` to zero at the beginning of `poll()` so I find it cleaner to initialise `revents` each time.

But in this code, I don't check `revents` since there can be only one possible event... so it does not make any differences.

File: pkg/container-collection/container-collection.go [hunk 2]
```diff
@@ -24,28 +24,78 @@ package containercollection
...
+	// been called, so that cc.containerEnrichers is fully set up.
+	for _, container := range cc.initialContainers {
+		for _, enricher := range cc.containerEnrichers {
+			enricher(container, true)
+		}
+		cc.containers.Store(container.Id, container)
+	}
+	cc.initialContainers = nil
+
+	cc.initialized = true
```
[the remark this answers is not in the document]
Comment (alban) [reply pkg/container-collection/container-collection.go:98]: I don't think it is worth adding complexity for a case which is clearly a bug. In Gadget Tracer Manager, `ContainerCollectionInitialize` is called from `newServer()` and no other routines can call `ContainerCollectionInitialize()` at the same time because nobody has a reference to the `GadgetTracerManager` yet. Same thing in `LocalGadgetManager`.

File: pkg/container-collection/options.go [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+			return
+		}
+		for _, container := range containers {
+			res, err := cc.dockercli.ContainerInspect(context.Background(), container.ID)
+			if err != nil {
+				continue
+			}
+			if res.State.Pid == 0 {
+				continue
+			}
```
[the remark this answers is not in the document]
Comment (alban) [reply pkg/container-collection/options.go:68]: It could be that the container has been deleted meanwhile. Not sure it's worth printing something in that case.

File: pkg/container-collection/options.go [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+				for _, m := range notif.ContainerConfig.Mounts {
+					mountSources = append(mountSources, m.Source)
+				}
+				containerDefinition := pb.ContainerDefinition{
+					Id:           notif.ContainerID,
+					Pid:          notif.ContainerPID,
+					MountSources: mountSources,
+				}
+
+				cc.AddContainer(&containerDefinition)
```
[the remark this answers is not in the document]
Comment (alban) [reply pkg/container-collection/options.go:195]: Here it cannot call methods from `GadgetTracerManager` because it might not have access to it...

In my next branch update, calling `GadgetTracerManager.AddContainer()` from here will not be necessary.

File: pkg/gadgettracermanager/gadgettracermanager.go [hunk 5]
```diff
@@ -450,7 +437,18 @@ func newServer(nodeName string, withPodInformer, withRuncFanotify, withBPF, with
 		tracers:  make(map[string]tracer),
 		withBPF:  withBPF,
 	}
-	g.ContainerCollectionInitialize()
+	opts := []containercollection.ContainerCollectionOption{
+		containercollection.WithCgroupEnrichment(),
+		containercollection.WithNamespaceEnrichment(),
```
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply pkg/gadgettracermanager/gadgettracermanager.go]: I made the comment in the wrong line. I think you're always using `WithKubernetesEnrichment` regardless the value of `withK8sClient`.

File: pkg/container-collection/container-collection.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply pkg/container-collection/container-collection.go:112]: > because nobody has a reference to the GadgetTracerManager yet. Same thing in LocalGadgetManager

This is true but you're supposing that the only users of `ContainerCollection` will always be `GadgetTracerManager` & `LocalGadgetManager.`We could also argue that calling multiple times  `ContainerCollectionInitialize` is a bug and we don't have to implement any measures against that (i.e. just drop the `initialized` member). 

I don't remember why we decided to embed `ContainerCollection` in `GadgetTracerManager`. If this would be a normal member then we could have a `containercollection.New(options ...ContainerCollectionOption) *ContainerCollection` and we won't have to worry about `ContainerCollectionInitialize` being called multiple times. Perhaps that's a cleaner approach?

File: pkg/container-collection/options.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply pkg/container-collection/options.go]: But if this is something else then the error won't be logged anywhere. I'd say it's better to print the logs now, if we found out they are to noise we can think how to improve them.

File: pkg/gadgettracermanager/gadgettracermanager.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (alban) [reply pkg/gadgettracermanager/gadgettracermanager.go]: This is fixed now :)

File: pkg/container-collection/container-collection.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (alban) [reply pkg/container-collection/container-collection.go:112]: > I don't remember why we decided to embed ContainerCollection in GadgetTracerManager.

I did it this way so that `GadgetTracerManager` implements the `Resolver` interface defined in `pkg/gadgets/interface.go`. This is so gadgets can call the different lookup methods such as `t.resolver.LookupMntnsByPod`.

But this could be changed to make it a normal `containercollection.New()` method as you suggest. We would just have to update the controller to something like that:

```diff
-factory.Initialize(tracerManager, mgr.GetClient())
+factory.Initialize(tracerManager.ContainerCollection, mgr.GetClient())
```

+ some other cleanups because ContainerCollection does not implement  `PublishEvent()` because it's not related to containers.

Should I do that in the same PR?

[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply pkg/container-collection/container-collection.go:112]: > Should I do that in the same PR?

I think it's perfectly fine to move to another PR. This has already enough changes :D

File: pkg/container-collection/container-collection.go [hunk 6]
```diff
@@ -24,28 +24,93 @@ package containercollection
...
+
+	// dockercli is the Docker API client to be used by
+	// WithDockerEnrichment
+	dockercli *dockerclient.Client
+
+	// client to talk to the k8s API server to get information about pods
+	k8sClient *kubernetes.Clientset
+
+	// runcNotifier to be used by WithRuncFanotify
+	runcNotifier *runcfanotify.RuncNotifier
```
Comment (mauriciovasquezbernal) [inline pkg/container-collection/container-collection.go:60]: I think we don't need to make these members. Couldn't they be declared in the scope of the different `ContainerCollectionOption` implementations?

**Review by mauriciovasquezbernal — COMMENTED** [review]
Some more comments.

File: pkg/container-collection/container-collection.go (hunk 6 above)
Comment (alban) [reply pkg/container-collection/container-collection.go]: Removed

File: pkg/container-collection/options.go [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+			return
+		}
+		cc.runcNotifier = runcNotifier
+
+		// Future containers
+		cc.containerEnrichers = append(cc.containerEnrichers, func(container *pb.ContainerDefinition, initial bool) {
+			if !initial {
+				return
+			}
+			cc.runcNotifier.WatchContainerTermination(container.Id, int(container.Pid))
```
[the remark this answers is not in the document]
Comment (alban) [reply pkg/container-collection/options.go]: Fixed. I changed the RuncNotifier method to remember the containers it is tracking, so it is safe to call the method for non-initial containers. It is also semantically more correct: it's not about whether it is an initial container, but whether the container is being added by RuncNotifier or by another component.

**Comment by alban** [comment]
@mauriciovasquezbernal I think I addressed all the comments. PTAL.

File: pkg/gadgettracermanager/k8s/k8s.go [hunk 8]
```diff
@@ -125,31 +125,14 @@ func (k *K8sClient) PodToContainers(pod *v1.Pod) []pb.ContainerDefinition {
...
-			log.Printf("Skip pod %s/%s: cannot find cgroup path: %v", pod.GetNamespace(), pod.GetName(), err)
-			continue
-		}
-		cgroupPathV2WithMountpoint, _ := containerutils.CgroupPathV2AddMountpoint(cgroupPathV2)
-		cgroupId, _ := containerutils.GetCgroupID(cgroupPathV2WithMountpoint)
-		mntns, err := containerutils.GetMntNs(pid)
-		if err != nil {
-			log.Printf("Skip pod %s/%s: cannot find mnt namespace: %v", pod.GetNamespace(), pod.GetName(), err)
-			continue
-		}
```
Comment (mauriciovasquezbernal) [inline pkg/gadgettracermanager/k8s/k8s.go:221]: Note to myself: Perhaps we could also remove the `pid, err := k.criClient.PidFromContainerId(s.ContainerID)` call above and have a pid enricher. (of course, something to be done in another pr)

**Review by mauriciovasquezbernal — APPROVED** [review]
Some very minor comments. Already LGTM! 

I did the following testing:
- podinformer
-- existing pods are considered
-- new created pods are considered
-- containers are removed when pod is deleted. (https://github.com/kinvolk/inspektor-gadget/issues/314 is still an issue but it's expected as this code is not rebased on top of https://github.com/kinvolk/inspektor-gadget/pull/317)

- fanotify 
-- existing containers are added (I suppose it works by using `WithInitialKubernetesContainers`. 
-- if an existing container is removed, then it's removed from the collection
-- new containers are added to the collection. 

I think there is still room for some more improvements and cleanups following this work, but it can be done in other PRs. Great work @alban

File: pkg/container-collection/options.go [hunk 9]
```diff
@@ -0,0 +1,386 @@
...
+						if _, ok := containerIDs[container.Id]; ok {
+							continue
+						}
+
+						// Make a copy instead of passing the same pointer at
+						// each iteration of the loop
+						newContainer := pb.ContainerDefinition{}
+						newContainer = container
+
+						cc.AddContainer(&newContainer)
```
[the remark this answers is not in the document]
Comment (alban) [reply pkg/container-collection/options.go:160]: I had this initially, and it adds the same container multiple times in the `sync.Map` (`gadgettracermanager -dump` would show the same container duplicated). This is because the same pointer is added multiple times, and at each iteration of the loop, the loop variable is updated, so that's updating the others containers already added in the `sync.Map`.
